base/reduce.jl

```diff
@@ -292,43 +292,9 @@ immutable Predicate{F}
 end
 (pred::Predicate)(x) = pred.f(x)::Bool
 
-const ShortCircuiting = Union{typeof(&), typeof(|)}
-
-## short-circuiting (sc) mapreduce definitions
-
-function mapreduce_sc_impl(f, op::typeof(&), itr)
-    for x in itr
-        f(x) || return false
-    end
-    return true
-end
-
-function mapreduce_sc_impl(f, op::typeof(|), itr)
-    for x in itr
-        f(x) && return true
-    end
-    return false
-end
-
-# mapreduce_sc tests if short-circuiting is safe;
-# if so, mapreduce_sc_impl is called. If it's not
-# safe, call mapreduce_no_sc, which redirects to
-# non-short-circuiting definitions.
-
-mapreduce_no_sc(f, op, itr::Any)           =  mapfoldl(f, op, itr)
-mapreduce_no_sc(f, op, itr::AbstractArray) = _mapreduce(f, op, itr)
-
-mapreduce_sc(f::Function,  op, itr) = mapreduce_no_sc(f, op, itr)
-mapreduce_sc(f::Predicate, op, itr) = mapreduce_sc_impl(f, op, itr)
-
-mapreduce_sc(f::typeof(identity), op, itr) =
-    eltype(itr) <: Bool ?
-        mapreduce_sc_impl(f, op, itr) :
-        mapreduce_no_sc(f, op, itr)
-
-mapreduce(f, op::ShortCircuiting, n::Number) = n
-mapreduce(f, op::ShortCircuiting, itr::AbstractArray) = mapreduce_sc(f,op,itr)
-mapreduce(f, op::ShortCircuiting, itr::Any)           = mapreduce_sc(f,op,itr)
+mapreduce(f, op::Union{typeof(&), typeof(|)}, n::Number) = n
```

Comment:
This actually should be 
```julia
mapreduce(f, op::Union{typeof(&), typeof(|)}, n::Number) = f(n)
                                                           ^^^^
```
I just now noticed this dumdum mistake of mine (thankfully calling `mapreduce(f,  rand([&,|]), x::Number)` instead of `f(x)` never took off, it had all the potential to be the new `x |> f`)

But maybe you should just remove all short-circuiting `mapreduce` methods all together for now, since the only thing using Predicate is `any`/`all` and nothing else will short-circuit. You could also to test to see if the`Predicate` thing still has any performance advantages and remove that as well if it doesn't.

Or you could just focus on the `any`/`all` part and leave me to clean up after my own mess in a followup PR, all valid options in my book :P